### PR TITLE
Start using remote_storage in S3 scrubber for PurgeGarbage

### PR DESCRIPTION
**Pull request by arpad-m**
Starts using the `remote_storage` crate in the S3 scrubber for the `PurgeGarbage` subcommand.

The `remote_storage` crate is generic over various backends and thus using it gives us the ability to run the scrubber against all supported backends.

Start with the `PurgeGarbage` subcommand as it doesn't use `stream_tenants`.

Part of #7547.

### CODE REVIEW

**Comment by arpad-m** [comment]
Marking this as a draft as it is untested.

**Comment by jcsp** [comment]
> Start with the PurgeGarbage subcommand as it doesn't use stream_tenants.

If you could implement a streaming listing in GenericRemoteStorage to enable this, it would also be useful in places like tenant deletion.  I'm not sure how hard it is to do that in a trait, as the version I have in the scrubber (`stream_listing`) uses `-> impl Stream<Item = anyhow::Result<ObjectIdentifier>> + 'a `, so the actual stream type will be different for different trait impls.

**Comment by arpad-m** [comment]
>  the actual stream type will be different for different trait impls

impl trait in trait has become stable together with async fn in trait, so it's available to users. For GenericRemoteStorage, it will likely have to return an enum though because it can only return one type.

File: storage_scrubber/src/garbage.rs [hunk 1]
```diff
@@ -324,41 +322,45 @@ impl std::fmt::Display for PurgeMode {
...
     // TODO: apply extra validation based on object modification time.  Don't purge
     // timelines whose index_part.json has been touched recently.
 
-    // Remove delimiter, so that object listing lists all keys in the prefix and not just
-    // common prefixes.
-    timeline_root.delimiter = String::new();
-    let key_stream = stream_listing(s3_client, &timeline_root);
-
-    key_stream.try_collect().await
+    let list = s3_client
```
Comment (skyzh) [inline storage_scrubber/src/garbage.rs:355]: s3_client -> storage_client in the future?

**Review by skyzh — APPROVED** [review]

File: storage_scrubber/src/garbage.rs (hunk 1 above)
Comment (arpad-m) [reply storage_scrubber/src/garbage.rs:355]: yeah, let's do this in a followup

**Pull request merged**